### PR TITLE
feat: add `--decay` option to highlight newer files

### LINKED ISSUE #486: feat:  Use font dimming to communicate modification/creation time
Porting ogham/exa#1222. Original text below...

--------------------

The title is probably a terrible description of the idea, but best I could do in limited space. Let me try to unpack it:

What I'm referring to is exactly what the project `k` does, where it calls this feature "rot" (gross to me, but I digress). [Here is a visual example.](https://github.com/supercrabtree/k#rotting-dates)

In prose: The font color of the date/time beside each file - be it creation, modification, or whatever - should be dimmer for dates that are older, and brighter for dates that are newer. This would give an immediate visual indication of which files are more important/relevant for the user at the present time.

I could see the dimming function coming in two modes, either of which would be very useful:

1. absolute (e.g., if all files displayed were modified either today or yesterday, all should be bright; only files modified a while ago should be dim)
2. relative to only the files displayed (e.g., regardless of absolute distance into the past, the least-recently-modified file would be dim, and the most-recently-modified file would be bright)

Ideally, the ability to toggle between these modes, or at least to set your preferred mode via an environment variable, would be super awesome. Not sure how feasible that would be to implement though.

For terminals which support the 256-color scheme, or even more, this should be doable regardless of hue, so not just black/white/gray.

Project `k` is written in shell code whereas `exa` is in Rust. But hopefully the former could still provide some useful ideas for implementation.

Anyway, curious to hear your thoughts of usability and feasibility!

### PR DESCRIPTION
**Pull request by aashirbadb**
This PR adds `--decay` option that highlights newer files. It has two modes:
-   `absolute`: Evaluates based on file modification time relative to the past year.
-   `relative`: Considers file modification time in relation to other files.

Both modes use a luminance formula `min + (1.0 - min) * exp(-4.0 * (1.0 - x))`, where `min` can be customized by using the 'EZA_MIN_LUMINANCE' environment variable.

Screenshots:
- relative mode
image
- absolute mode
image
- tree
image

(I'm calling it `decay` because I couldn't think of a better name 😅. I'd greatly appreciate any suggestions for a more fitting name)

Resolves #486

### CODE REVIEW

**Comment by PThorpe92** [comment]
I like the term decay.. it's a hell of a lot better than `rot` lmao

**Review by cafkafk — CHANGES_REQUESTED** [review]
This is looking cool.

For conventional commits: `doc: add --decay option and EZA_MIN_LUMINANCE env var` needs to be `docs: add --decay option and EZA_MIN_LUMINANCE env var`

**Review by cafkafk — CHANGES_REQUESTED** [review]
Could we have default behavior where `--decay` without any arguments default to `--decay relative`?

**Comment by aashirbadb** [comment]
> Could we have default behavior where `--decay` without any arguments default to `--decay relative`?

I tried adding it but it looks like `TakesValue::Optional` doesn't work like it should.
Passing `--decay something` gives `(decay, Some(something))` instead of `(decay, Some(default_value))`(if something is invalid option for the flag) or `(decay, Some(something))` (if something is valid option).

https://github.com/eza-community/eza/blob/9805dd587936ba9e8e61d216b5f96b51b4676a24/src/options/parser.rs#L201-L207

Should I fix this(maybe in a separate PR?) or wait for  #197 and implement it then?

**Comment by PThorpe92** [comment]
>Should I fix this(maybe in a separate PR?) or wait for https://github.com/eza-community/eza/pull/197 and implement it then?

I am currently working through another implementation that gives a flag a default value. Once I'm finished I'll take a look and see if we can get it figured out.

Hopefully clap will be very soon. Honestly, this is a really exciting change and part of me does want to wait and release the functionality with `v1.0`

Feel free to come to the matrix channel too if you want to talk about it

**Comment by cohml** [comment]
> Feel free to come to the matrix channel too if you want to talk about it

Link to said channel please? I'm not familiar with this "matrix", but would like to be a fly on the wall. Thanks.

**Comment by PThorpe92** [comment]
If you go to the main page the "Chat on gitter" button

https://matrix.to/#/#eza-community:gitter.im is the link

**Comment by Delapouite** [comment]
Also, should the option `--color-scale` be renamed to avoid any confusion about where it applies ? `size` vs `date`?

**Comment by PThorpe92** [comment]
also, I was able to get the `--icons` flag to take a default value and have a `TakesValue::Optional(Some(VALUE))`

So if you want to have this take an optional argument and set a default value, you can look at how icons does it now.

**Comment by Delapouite** [comment]
You're right PThorpe92, the notion of `color-scale` could even reach further.

Non exhaustive list of possible columns :

- date: old → new (this PR)
- size: small → big (kind of current behavior of `--color-scale`, but limited to magnitude kilo, mega, tera…)
- links count : few → lot
- octal permissions : very-restricted → open bar (777)
- …

**Comment by PThorpe92** [comment]
I think the way to go would be to have breaking changes, and make sure that --color-scale has a value. It will break people's alias's, but it will also alert them to the new feature.

Have you tested this alongside the current --color-scale flag? How does that work?

**Comment by aashirbadb** [comment]
> That is a good question.. how would you feel about making this a value for the color-scale flag? e.g. `--color-scale=age` || `--color-scale=size`

`--decay` takes two values so I can just change it so that `relative` is default value when `--color-scale=age` is used and it can be changed to absolute mode using `--color-scale=age --decay=absolute`.

> I think the way to go would be to have breaking changes, and make sure that --color-scale has a value. It will break people's alias's, but it will also alert them to the new feature.

We can also allow multiple choices using comma separated list(`--color-scale=age,size`) and maybe add an `all` choice which could be optional flag to not introduce a breaking change?

**Comment by PThorpe92** [comment]
My thought process was kinda specifically to introduce the feature. Adding a value to a flag isn't nearly as exciting as the new `decay` option. But I do like that idea. It could just default to `all`

**Comment by aashirbadb** [comment]
I think it would be better to remove the `--decay` flag and use  `--color-scale` and add a new flag(something like `--color-scale-mode`) to select the mode(`fixed`,`gradient`). 

Fixed would work like the current `--color-scale` and gradient would work like relative mode. 

This would also make it easier to add more columns like Delapouite suggested.

**Comment by aashirbadb** [comment]
I added this to size and it looks pretty cool.

image

**Review by cafkafk — CHANGES_REQUESTED** [review]
Looks like you have a formatting issue, if you're using nix, this is easily fixed with `nix fmt`.

Also, your breaking change should have a `BREAKING` footer (or w/e exactly it's called) as specified by conventional commits, you can find more info about it in the conventional commits spec.

File: README.md [hunk 1]
```diff
@@ -90,7 +90,8 @@ eza’s options are almost, but not quite, entirely unlike `ls`’s.
 - **-x**, **--across**: sort the grid across, rather than downwards
 - **-F**, **--classify**: display type indicator by file names
 - **--colo[u]r=(when)**: when to use terminal colours (always, auto, never)
-- **--colo[u]r-scale**: highlight levels of file sizes distinctly
+- **--colo[u]r-scale=(field)**: highlight levels of `field` distinctly(all, age, size)
```
Comment (MartinFillon) [inline README.md:93]: I don't like age as `Field name` could you use the name used on the `--header` option

File: src/options/theme.rs [hunk 2]
```diff
@@ -204,13 +192,13 @@ mod terminal_test {
...
+    // test!(scale_4:  ColourScale <- [                                 ];   Last => Ok(ColourScale::Fixed));
 
-    test!(scale_5:  ColourScale <- ["--color-scale", "--colour-scale"];   Complain => err OptionsError::Duplicate(Flag::Long("color-scale"),  Flag::Long("colour-scale")));
-    test!(scale_6:  ColourScale <- ["--color-scale",                 ];   Complain => Ok(ColourScale::Gradient));
-    test!(scale_7:  ColourScale <- [                 "--colour-scale"];   Complain => Ok(ColourScale::Gradient));
-    test!(scale_8:  ColourScale <- [                                 ];   Complain => Ok(ColourScale::Fixed));
+    // test!(scale_5:  ColourScale <- ["--color-scale", "--colour-scale"];   Complain => err OptionsError::Duplicate(Flag::Long("color-scale"),  Flag::Long("colour-scale")));
+    // test!(scale_6:  ColourScale <- ["--color-scale",                 ];   Complain => Ok(ColourScale::Gradient));
+    // test!(scale_7:  ColourScale <- [                 "--colour-scale"];   Complain => Ok(ColourScale::Gradient));
+    // test!(scale_8:  ColourScale <- [                                 ];   Complain => Ok(ColourScale::Fixed));
```
Comment (MartinFillon) [inline src/options/theme.rs:203]: Please remove those comments and test the code

Comment (MartinFillon) [reply src/options/theme.rs:203]: If not possible to test just remove it

File: src/output/decay.rs [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+fn update_information_recursively(
+    information: &mut ColorScaleInformation,
+    files: &[File<'_>],
+    dot_filter: DotFilter,
+    git: Option<&GitCache>,
+    git_ignoring: bool,
+    depth: TreeDepth,
+    r: Option<RecurseOptions>,
+) {
```
Comment (MartinFillon) [inline src/output/decay.rs:112]: Why is this function not in the impl above ?

File: src/output/render/size.rs [hunk 4]
```diff
@@ -21,28 +23,49 @@ impl f::Size {
...
             SizeFormat::DecimalBytes  => NumberPrefix::decimal(size as f64),
             SizeFormat::BinaryBytes   => NumberPrefix::binary(size as f64),
             SizeFormat::JustBytes     => {
-
                 // Use the binary prefix to select a style.
                 let prefix = match NumberPrefix::binary(size as f64) {
-                    NumberPrefix::Standalone(_)   => None,
-                    NumberPrefix::Prefixed(p, _)  => Some(p),
+                    NumberPrefix::Standalone(_) => None,
+                    NumberPrefix::Prefixed(p, _) => Some(p),
```
Comment (MartinFillon) [inline src/output/render/size.rs:38]: please format those lines

**Review by MartinFillon — CHANGES_REQUESTED** [review]

**Pull request closed without merging**